[stdlib] make `_overrideLifetime()` functions public

glessard opened the pull request:
These functions are discussed in the lifetime annotations pitch: https://github.com/swiftlang/swift-evolution/pull/2750
They are necessary to implement functions returning non-escapable values, such as `Data.bytes`.

Addresses rdar://150400414

glessard committed:
[stdlib] make `_overrideLifetime` functions public

These are discussed in the lifetime annotations pitch: https://github.com/swiftlang/swift-evolution/pull/2750

Addresses rdar://150400414

glessard committed:
[test] use the stdlib _overrideLifetime functions

glessard commented:
@swift-ci please smoke test

meg-gupta reviewed (approved):
lgtm

atrick reviewed (approved):
Thanks! This is super helpful.

glessard committed:
[test] remove more copies of `_overrideLifetime()`

atrick commented:
@glessard you can remove these TODOs now:
```
  // TODO: Remove @_unsafeNonescapableResult. Instead, the unsafe dependence
  // should be expressed by a builtin that is hidden within the function body.
```
They were originally to discourage people from using `@_unsafeNonescapableResult` in more places, but we won't be able to remove that annotation any time soon and the TODO is unhelpful for anyone not working on the compiler.

glessard committed:
[stdlib] remove TODOs

These TODOs aren’t particularly actionable. What we really want is a way to define `_overrideLifetime()` in a not-unsafe way, and that will probably be a `Builtin` operation.

glessard commented:
@swift-ci please smoke test
